### PR TITLE
test(ts): change-based selective integration testing

### PR DESCRIPTION
**Pull request by yonib05**
## What

Adds change-based selective integration testing for the TypeScript SDK. Instead of always running the full `integ-node` + `integ-browser` suite, CI and local runs now execute only the integration specs whose module graph depends on the files a change touched, with a fail-safe fallback to the full suite for structural changes.

## Why

The full integration suite hits live AWS/Bedrock and is slow. For a localized source fix, most of that suite is irrelevant. This narrows the run to the covering specs using Vitest's native module-graph tracing (`vitest related`), which resolves the `$/sdk` path alias correctly. No new dependencies and no new orchestration tooling.

## How it works

`test-infra/scripts/run-selective-ts.sh` classifies the diff against the base ref into three branches:

1. **Structural change** (dependency manifests, lockfiles, any `strands-ts` tsconfig, `vitest.config.ts`, shared `test/integ/__fixtures__/` and `__resources__/`, `strandly/`, the script itself, or TypeScript CI workflows) runs the full suite.
2. **No TypeScript source change** skips the integration job.
3. **Otherwise** passes the changed source files to `vitest related --project integ-node --project integ-browser`.

The same script backs both the local `npm run test:integ:selective` command and the CI step, so local and CI scope match for a given change. Locally it diffs the working tree against the merge-base (so uncommitted edits are included); in CI it diffs against the PR base SHA, falling back to the merge-queue base on `merge_group` events.

## Fail-safe design

Every uncertain path runs the full suite rather than skipping: unresolvable base ref, failed diff, or any change the static graph cannot trace (binary assets imported via Vite `?url`, tsconfig path-alias definitions, wasm/wit inputs) is routed to the structural fallback. The goal is never to skip a test that should run.

## Scope

TypeScript only. Python selective testing is intentionally deferred to a follow-up because its options (pytest-testmon and friends) carry real trade-offs with nondeterministic AWS-backed tests and stateful coverage data that warrant separate evaluation.

## Files

- `test-infra/scripts/run-selective-ts.sh` (new): the selection logic
- `package.json`: `test:integ:selective` script
- `.github/workflows/typescript-integration-test.yml`: wires the script into the existing integration-test workflow (full git history for diffing; selective run with the PR/merge-queue base SHA)
- `CONTRIBUTING.md`: local usage docs

## Verification

All branch classifications were verified via a dry-run mode that prints the chosen branch without executing tests (so no live AWS calls): structural cases (tsconfigs, `__resources__` assets, `strandly/`) route to the full suite, a normal source edit narrows to the selective set, and unrelated changes skip.

## Note for reviewers

Integration tests still run against live AWS/Bedrock. This change reduces how many run for a localized change; it does not change per-test reliability.

### CODE REVIEW

File: test-infra/scripts/run-selective-ts.sh [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+echo "$TS_SOURCE" | sed 's/^/  /'
+[[ -n "$DRY_RUN" ]] && exit 0
+# Collect paths into an array (relative to strands-ts/) so filenames with
+# spaces survive. while-read keeps this portable to macOS bash 3.2 (mapfile
+# is bash 4+). TS_SOURCE is guaranteed non-empty by the Branch 2 check above.
+files=()
+while IFS= read -r f; do
+  [[ -n "$f" ]] && files+=("$f")
+done < <(echo "$TS_SOURCE" | sed -E 's#^(strands-ts|strands-wasm|wit)/##')
+( cd strands-ts && npx vitest related "${files[@]}" \
```
[the remark this answers is not in the document]
Comment (yonib05) [reply test-infra/scripts/run-selective-ts.sh:127]: Fixed in 6750edca. Added `'@strands-agents/sdk' -> path.resolve(__dirname, "./src")` to the alias map of both integ projects in `vitest.config.ts`, so `vitest related` resolves the package specifier (and its subpaths) to source and can trace `src/` edits to it.

Verified empirically: before the fix, editing `src/agent/agent.ts` selected 0 of the 9 package-only importers; after, it correctly selects `agent.cancel.test.ts`, `interrupt.test.ts`, `agent-as-tool.test.ts`, the conversation-manager, mcp, and http-request tests. Selectivity is preserved (a leaf-file edit still narrows to 3 of 40, not the full suite).

Heads-up for maintainers: this means the package-importing integ tests now resolve against `src` rather than the built `dist`. The packaged artifact remains covered by the separate `test:package` job. If integ coverage of the built package is intended, flag it and I will instead route `src` changes more conservatively.

File: test-infra/scripts/run-selective-ts.sh [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    d=$(git rev-list --count "$ref"..HEAD 2>/dev/null || echo 999999)
+    if [[ "$d" -lt "$best" ]]; then BASE="$ref"; best="$d"; fi
+  done
+fi
+
+# --- Compute changed files ---
+# merge-base diff INCLUDING uncommitted working-tree changes, so the local
+# inner loop tests what you just edited. (Two-arg form: base..working-tree.)
+MERGE_BASE="$(git merge-base "$BASE" HEAD 2>/dev/null || echo "$BASE")"
+CHANGED="$(git diff --name-only "$MERGE_BASE" 2>/dev/null)" || {
```
[the remark this answers is not in the document]
Comment (yonib05) [reply test-infra/scripts/run-selective-ts.sh]: Fixed in 6750edca. Appended `git ls-files --others --exclude-standard` to the changed set, so brand-new untracked source files now select their covering tests. `--exclude-standard` honours `.gitignore`, and on CI (HEAD == base SHA) there are no local edits so it reduces to the committed diff. The "including uncommitted edits" wording is now accurate rather than softened.

Verified: a new untracked `src/` file routes to a selective run and is listed; an untracked docs file still skips; a clean tree reports no changes.

File: test-infra/scripts/run-selective-ts.sh [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+# dependencies of a test, so they must fail safe:
+#   - dependency manifests / lockfiles
+#   - any tsconfig under strands-ts (nested too: src/, test/integ/ define the $/sdk alias)
+#   - the vitest config
+#   - shared test fixtures AND binary resources imported via Vite `?url`
+#     (the module graph does not traverse `?url` edges in reverse)
+#   - strandly/ (workspace member that CI triggers on but the graph can't trace)
+#   - this orchestration script itself
+#   - the TypeScript CI workflows
+STRUCTURAL='^package\.json$|^package-lock\.json$|^strands-ts/package\.json$|^strands-ts/(.*/)?tsconfig.*\.json$|^strands-ts/vitest\.config\.ts$|^strands-ts/test/integ/__fixtures__/|^strands-ts/test/integ/__resources__/|^strandly/|^test-infra/scripts/run-selective-ts\.sh$|^\.github/workflows/typescript-'
```
[the remark this answers is not in the document]
Comment (yonib05) [reply test-infra/scripts/run-selective-ts.sh]: Done in 0910064a. Added a `SELECTIVE_CHANGED_FILES` test seam (overrides the git-derived change set when defined) and `test-infra/scripts/test/classify.test.sh`: 21 pure-bash cases that feed synthetic file lists through the structural/skip/selective classifier with no git state, network, or live AWS. It locks in the full structural set (lockfiles, nested tsconfigs, `__fixtures__`/`__resources__`, `strandly/`, the script itself, TS workflows) so a future regex edit cannot silently start skipping tests. It already earned its keep by catching an empty-change-set edge case during development.

File: test-infra/scripts/run-selective-ts.sh [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+TS_SOURCE="$(echo "$CHANGED" | grep -E '^strands-ts/' || true)"
+if [[ -z "$TS_SOURCE" ]]; then
+  echo "No strands-ts source changes — skipping integration tests."
+  exit 0
+fi
+
+# --- Branch 3: selective ---
+# Pass changed source files to Vitest's module-graph tracer, scoped to both
+# integ projects. vitest related exits 0 with "No test files found" when none
+# depend on the changes — a valid skip.
```
[the remark this answers is not in the document]
Comment (yonib05) [reply test-infra/scripts/run-selective-ts.sh:110]: Added an explicit comment in 42f26efe pinning this. The selective branch now calls out that the fail-safe invariant depends on vitest v4 exiting 0 (not non-zero) on the "no covering spec" case, ties it to the version floor in `strands-ts/package.json`, and says a future major bump that changes this must revisit the branch.

Went with the comment rather than a `vitest related` smoke test on purpose: `classify.test.sh` is deliberately hermetic (no git, no network, no vitest), and adding a vitest-invoking case would undercut that. Resolving.

**Review by chaynabors — APPROVED** [review]
